### ISSUE TITLE
[Feature]: Model confidence calibration - ECE metric + temperature scaling

### ISSUE THREAD

**Issue by Adit-Jain-srm** [Adit-Jain-srm/NightmareNet#215]
## Problem

NightmareNet's evaluation reports accuracy and robustness (correct predictions under distortion) but does NOT assess whether the model's confidence is reliable. A model that outputs 95% confidence on a wrong answer under adversarial attack is more dangerous than one that says "I'm uncertain" - because downstream systems trust high-confidence predictions.

Current evaluation gap:
- No measure of whether confidence correlates with correctness (calibration)
- No detection of when confidence drops under distortion (uncertainty degradation)
- No option to abstain when the model is uncertain (selective prediction)

Paper Section 3.2 mentions the EU AI Act Article 15 requirement for "appropriate level of accuracy, robustness, and cybersecurity" - calibration is a core component of reliability that regulators increasingly expect.

**Relevant files:**
- `nightmarenet/evaluation/evaluator.py` (extends `evaluate()` with calibration metrics)
- `nightmarenet/evaluation/metrics.py` (new metric functions)
- `configs/default.yaml` (add `evaluation.calibration` config block)

### Phase 1: ECE metric + Temperature Scaling (this issue)

1. **`nightmarenet/evaluation/calibration.py`** - New module:
   - `compute_ece(confidences, predictions, labels, n_bins=15)` - Expected Calibration Error
   - `TemperatureScaler.fit(logits, labels)` - Learn optimal temperature T on validation set
   - `TemperatureScaler.calibrate(logits)` - Apply learned temperature to soften/sharpen predictions
   - `reliability_diagram_data(confidences, labels, n_bins)` - Bin data for plotting

2. **Integration with Evaluator:**
   - When `evaluation.calibration.enabled: true`, compute ECE before and after temperature scaling
   - Report in comparison dict: `ece_before`, `ece_after`, `optimal_temperature`
   - Include in markdown report

3. **Config:**
   ```yaml
   evaluation:
     calibration:
       enabled: false
       ece_bins: 15
       temperature_scaling: true
   ```

### Phase 2: Abstention / Selective Prediction (follow-up issue)

Selective prediction (abstain when confidence < threshold) is a separate concern that builds on calibration. Will open as a follow-up after Phase 1 merges.

## Acceptance Criteria

- [ ] `nightmarenet/evaluation/calibration.py` exists with `compute_ece()` and `TemperatureScaler`
- [ ] ECE formula correct: weighted average of |accuracy - confidence| across bins
- [ ] Temperature scaling: single scalar T, learned via NLL minimization on validation logits
- [ ] ECE computed on classification models (requires softmax probabilities, not logits)
- [ ] Integration: `Evaluator.evaluate()` includes calibration metrics when enabled
- [ ] Config: `evaluation.calibration.enabled` (default false, opt-in)
- [ ] Reliability diagram data exposed in JSON (bin_confidences, bin_accuracies, bin_counts)
- [ ] `nightmarenet evaluate` CLI output includes ECE when enabled
- [ ] Test: perfectly calibrated model → ECE ≈ 0
- [ ] Test: overconfident model → ECE > 0, temperature T > 1
- [ ] Test: temperature scaling reduces ECE
- [ ] No regression when `calibration.enabled: false` (default path unchanged)

## Technical Notes

- **ECE is computed on probabilities (after softmax), not logits.** Common implementation bug.
- **Temperature T is a single scalar** optimized via `torch.optim.LBFGS` on the NLL loss over the validation set. Converges in < 1 second for typical datasets.
- **Calibration set must be SEPARATE from test set.** Use the existing `eval_split_ratio` mechanism (the held-out evaluation split) as the calibration set.
- **Reference:** Guo et al. 2017 "On Calibration of Modern Neural Networks" (all deep nets are miscalibrated; temperature scaling is the simplest effective fix).

## Scope

Medium-Large. ~4-5 hours. ECE implementation + temperature scaling + evaluator integration + tests.

---

> **Important:** Please read [CONTRIBUTING.md](https://github.com/Adit-Jain-srm/NightmareNet/blob/main/CONTRIBUTING.md) in full before requesting assignment. Assignment requests without a technical approach will not be considered. PRs demonstrating thorough research, benchmarks, or creative approaches beyond minimum criteria may earn bonus XP labels.

**Comment by Aryanbuha890** [comment]
Hi @Adit-Jain-srm,

I would like to be assigned to this issue. Below is my detailed technical approach for implementing ECE and Temperature Scaling:

### Technical Approach

1. **Calibration Computations (`nightmarenet/evaluation/calibration.py`)**
   - **ECE Calculation (`compute_ece`)**: I will implement ECE by grouping predicted probabilities into $M$ equally-spaced bins (default 15). For each bin $B_m$, I will calculate average confidence and accuracy, then sum the absolute differences weighted by the bin size relative to total samples:
     $$\text{ECE} = \sum_{m=1}^{M} \frac{|B_m|}{N} |\text{acc}(B_m) - \text{conf}(B_m)|$$
     To ensure correctness, ECE will be computed strictly on post-softmax probabilities.
   - **Reliability Diagram Data**: Provide a helper to output bin counts, average confidence, and average accuracy as JSON-serializable list structures.

2. **Temperature Scaling Optimization**
   - **`TemperatureScaler`**: I will implement Temperature Scaling by optimizing a single scalar parameter $T$ (initialized at 1.0) using PyTorch. 
   - **Optimization Method**: Logits will be scaled as $z / T$. We will optimize $T$ by minimizing the negative log-likelihood (CrossEntropyLoss) using `torch.optim.LBFGS` on a validation split, which converges quickly and robustly.

3. **Evaluator & CLI Integration**
   - I will integrate this block in `Evaluator.evaluate()` behind the configuration flag `evaluation.calibration.enabled`.
   - If enabled, the evaluator will use the existing `eval_split_ratio` to hold out a dedicated calibration set separate from the final test set, optimize temperature $T$, and then report `ece_before`, `ece_after`, and `optimal_temperature` in the final summary.

4. **Testing Criteria**
   - Add unit tests verifying:
     - Synthetic data checks: perfect predictions yield ECE close to 0; overconfident models yield high ECE and optimized $T > 1$.
     - Temperature scaling successfully reduces ECE on miscalibrated synthetic validation inputs.

Please assign this issue to me!

**Comment by Adit-Jain-srm** [comment]
Assigned @Aryanbuha890
